### PR TITLE
#1357

### PR DESCRIPTION
**Pull request by tfhuhtal**
### **Pull Request: Add functionality to import students from course realisations to organisation surveys**  

**Issue Addressed**: [#1357](https://github.com/UniversityOfHelsinkiCS/palaute/issues/1357)  

This PR implements the ability to import students from course realisations into the organisation survey editor. The following changes and features have been added:  

---

### **Changes and Features**  

1. **Import Functionality**  
   - Added an input field in the organisation survey editor where users can input course codes.  
   - Users can select one or more courses from which to import students.  
   -  Users can search courses based on year and semester.

2. **Dynamic Updates**  
   - Organisation survey editor now dynamically updates the student count based on the imported students from the selected course realisations.  

3. **Localization**  
   - Added locale support for new strings to ensure translations align with the project's standards.  

4. **Testing**  
   - Added comprehensive tests to cover the new functionality, including:  
     - Importing students from valid courses.  
     - Ensuring UI reflects the updated state after import operations.

### CODE REVIEW

File: cypress/integration/organisationSurvey.spec.js
Comment (HRemonen) [inline cypress/integration/organisationSurvey.spec.js:1]: Curious why did you change the formik-date-picker-field-startDate/endDate data-cy's to formik-date-picker-field-startDate/endDate-inputs

I know for a fact that both of these are there, the input is only the input field, but the former is the whole component.

Also you could update the other organisation survey tests to include the new courseIds array, even if empty

File: src/client/components/common/YearSemesterSelector.jsx
Comment (HRemonen) [inline src/client/components/common/YearSemesterSelector.jsx:1]: It seems like the new prop `organisationSurvey` isn't used, so let's remove it from the code

File: src/client/pages/FeedbackTarget/Edit/EditOrganisationSurvey.jsx
Comment (HRemonen) [inline src/client/pages/FeedbackTarget/Edit/EditOrganisationSurvey.jsx:1]: We have a [queryClient](https://github.com/UniversityOfHelsinkiCS/palaute/blob/master/src/client/util/queryClient.js#L3) utility function, which should be used instead of importing the acual queryClient. But I would remove this invalidation from here completely and move it to the [editMutation](https://github.com/UniversityOfHelsinkiCS/palaute/blob/c39e3f51551299bd45170a0009a2695c8c843a61/src/client/pages/Organisation/useOrganisationSurveyMutation.js#L31), the same way the organisationSurveys are invalidated there.

Why do we want to invalidate the fbt?

File: src/client/pages/Organisation/CourseSearchInput.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+          onChange={(_, courses) => {
+            formikProps.setFieldValue('courses', courses)
+          }}
+          inputValue={courseSearch}
+          onInputChange={(_, value) => {
+            setCourseSearch(value)
+          }}
+          options={data ?? []}
+          getOptionLabel={getOptionLabel}
+          renderInput={(params: any) => (
```
Comment (HRemonen) [inline src/client/pages/Organisation/CourseSearchInput.tsx:69]: If you can, it's always benefitial to type this params also :)

File: src/client/pages/Organisation/CourseSearchInput.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+interface InitialValues {
+  name: Locales
+  startDate: Date
+  endDate: Date
+  studentNumbers: []
+  teachers: []
+  courses: []
+}
+
+type OrganisationCode = any
```
Comment (HRemonen) [inline src/client/pages/Organisation/CourseSearchInput.tsx:23]: Can we use a stricter type for the OrganisatioCode here?

File: src/client/pages/Organisation/CourseSearchInput.tsx
Comment (HRemonen) [inline src/client/pages/Organisation/CourseSearchInput.tsx:1]: Good job, a few suggestions thought:
- We have all kinds of utility functions in Norppa, for example for getting [date strings](https://github.com/UniversityOfHelsinkiCS/palaute/blob/c39e3f51551299bd45170a0009a2695c8c843a61/src/client/pages/Organisation/OrganisationSurveyItem.jsx#L46) and the [localised value](https://github.com/UniversityOfHelsinkiCS/palaute/blob/c39e3f51551299bd45170a0009a2695c8c843a61/src/client/pages/Organisation/OrganisationSurveyItem.jsx#L51) such as the name.

File: src/server/services/organisations/organisationSurveys.js [hunk 3]
```diff
@@ -444,6 +444,37 @@ const deleteOrganisationSurvey = async feedbackTargetId => {
   }
 }
 
+const getStudentNumbersFromCourseIds = async courseIds => {
+  try {
+    if (!courseIds || courseIds.length < 1) return []
```
Comment (HRemonen) [inline src/server/services/organisations/organisationSurveys.js:449]: Small, but I think it's better to check if courseIds.length === 0 istead of < 1. Easier to read

File: src/server/routes/organisations/organisationSurveyController.js [hunk 4]
```diff
@@ -56,24 +57,39 @@ const getOrganisationSurveys = async (req, res) => {
...
   })
 
   if (!hasAdminAccess) throw new ApplicationError('Only organisation admins can create organisation surveys', 403)
 
+  const studentNumbersFromCourseIds = await getStudentNumbersFromCourseIds(initialCourseIds)
+
   // Remove duplicates from studentNumbers and teacherIds
-  const studentNumbers = [...new Set(initialStudentNumbers)]
+  const studentNumbers = [
+    ...new Set([...initialStudentNumbers, ...studentNumbersFromCourseIds.filter(n => n !== null)]),
```
Comment (HRemonen) [inline src/server/routes/organisations/organisationSurveyController.js:79]: You can use `...studentNumbersFromCourseIds.filter(n => Boolean(n))` or even `...studentNumbersFromCourseIds.filter(n => n)` here instead

File: src/server/routes/organisations/organisationSurveyController.js [hunk 5]
```diff
@@ -91,7 +107,13 @@ const editOrganisationSurvey = async (req, res) => {
   const { user, body } = req
   const { id } = req.params
 
-  const updates = _.pick(body, ['name', 'startDate', 'endDate', 'teacherIds', 'studentNumbers'])
+  const updates = _.pick(body, ['name', 'startDate', 'endDate', 'teacherIds', 'studentNumbers', 'courseIds'])
+
+  const studentNumbersFromCourseIds = await getStudentNumbersFromCourseIds(updates.courseIds)
+
+  updates.studentNumbers = [
+    ...new Set([...updates.studentNumbers, ...studentNumbersFromCourseIds.filter(n => n !== null)]),
```
Comment (HRemonen) [inline src/server/routes/organisations/organisationSurveyController.js:115]: Same filter trick as before, also if we are using the exact same thing here and there then maybe you should filter it in to a variable and use that to save some computation

File: src/client/pages/Organisation/CourseSearchInput.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    return `${course?.name[lang] || ''} (${startDate} - ${endDate})`
+  }
+
+  const handleDateRangeChange = (nextDateRange: DateRange) => {
+    setDateRange(nextDateRange)
+  }
+
+  return (
+    <>
+      <Grid xs={10} item>
```
Comment (HRemonen) [inline src/client/pages/Organisation/CourseSearchInput.tsx:54]: Let's remember to migrate this Grid to the MUI Grid2 after we merge with the master

**Review by HRemonen — CHANGES_REQUESTED** [review]
Good job, a few refinements and we are good to go I think!

File: src/client/pages/Organisation/CourseSearchInput.tsx (hunk 6 above)
Comment (tfhuhtal) [reply src/client/pages/Organisation/CourseSearchInput.tsx:54]: Resolving this later
